### ISSUE TITLE
Feature Request: cache syntax highlighting to improve 'relativenumber' scroll performance

### ISSUE THREAD

**Issue by bluz71** [vim/vim#1735]
Hello Vim team,

Spawning a new feature request from the existing [Ruby high CPU usage #282 issue](https://github.com/vim/vim/issues/282). Note, the [vim-ruby #243 issue](https://github.com/vim-ruby/vim-ruby/issues/243) is fundamentally the same issue as well.

Basically the issue boils down to syntax highlighting needing to be calculated for every visible line when scrolling up or down with *relativenumber* enabled. The more lines visible and the slower the host CPU the worse the performance hit especially for languages with complex syntax highlighters such as Ruby.

In issue [#282](https://github.com/vim/vim/issues/282#issuecomment-164013461) Bram said this (Dec 2015):

> When moving the cursor around, with 'relativenumber' or 'cursorcolumn'
> set, cause a full redraw, because every line changes. I'm wondering if
> it is possible to cache the result of syntax highlighting. It would
> require storing the "attr" value, before it's mixed with other
> attributes, such as from 'cursorcolumn'. So long as we don't scroll the
> screen update would be much faster. These days memory is plenty, thus
> that isn't a problem.
> 
> As soon as you page forward it's just as slow, of course, thus making
> the syntax highlighting faster is still desired.

I suspect, if implemented, that this suggestion indeed would greatly help scroll performance for those of us who enable *relativenumber*. A new setting possibly? *cachesyntax*?

I leave it to the Vim masters to either: do this, or not do this, or schedule later, or close if too hard.

Thank you.

**Comment by mgraham** [comment]
Ok, I don't know the internals of cursorline or vim's syntax highlighting, but it seems to me that it should be possible to completely separate cursorline highlighting from syntax highlighting?

Usually, the cursorline just changes the background colour for the current line.  So...

* when drawing the line the cursor is on, set the background colour to the highlight colour.
* when drawing any other line set the background colour to the regular background colour. 

I don't understand why the entire screen's syntax highlighting needs to be recalculated because the background colour of one line gets toggled.

**Comment by mgraham** [comment]
For instance, the screen's syntax highlighting does not seem to need to be recalculated every time a visual text selection changes.

**Comment by neoadventist** [comment]
Awesome! I've been wondering why my screen has been flickering every time I go to the next line. I would love to know how to fix this! @mgraham @bluz71

**Comment by bluz71** [comment]
@neoadventist, 

The enhancement needs to happen inside Vim & Neovim, so for now there is not much you can do except disable `relativenumber`.

**Comment by her** [comment]
If you comb through the various postings about this issue across the internet these suggestions come up frequently as a fix: 

```
set ttyfast
set lazyredraw
set regexpengine=1
```

However these I suspect are actually just fixing people's issues with 3rd party plugins.

The real fix for this is:

```
set norelativenumber
```

I think this feature is simply broken when applied to ruby filetypes due to what @bluz71 and @brammool already mentioned. 

Hopefully this helps others as it took me some time to find the culprit. Cheers.

**Comment by bluz71** [comment]
@her,

I agree with:

```viml
set ttyfast
set regexpengine=1
```

However,  enabling `lazyredraw` has lead to weird redraw and slow character deletion issues in my experience. I experimented with it quite a bit and in the end concluded that it is best left disabled.

I also recommend this setting:

```viml
set synmaxcol=200
```
Only syntax highlight the first 200 chars of a line. Very long lines are a killer with `relativenumber`. 

The Ruby syntax highlighter contains hundreds of regexps, multiple by the number of visible lines can equal very bad scroll performance with `relativenumber`. Caching syntax highlighting is my number one most wanted Vim enhancement. Good to see the first post of this issue now with 62 likes (and slowly growing).

**Comment by folofjc** [comment]
I have this issue and do not do pgp or ruby syntax highlighting. It also just started happening recently, so might have come from an update.

**Comment by bluz71** [comment]
I noticed that [#282](https://github.com/vim/vim/issues/282) has been closed due to patches  8.1.0372 - 8.1.0374

Commit [bd9a53c](https://github.com/vim/vim/commit/bd9a53c06c8869ad811cb3dd01a309c9be7d7a63) (patch 8.1.0374) implements a `relativenumber` optimisation, _only redraw the number column, not all lines_.

I have updated and tested, and indeed when scrolling within a page, with `relativenumber` enabled, navigating the cursor seems much smoother.

Note, scrolling off the top or bottom of the page, via long-pressing `j` or `k` may still be slow sometimes, but in that case every line on the page is changing AND syntax highlighting will be applied to lines coming into view. That was never the crux of this issue which was only about scrolling within a single static screen.

Thanks to the Vim maintainers, very much appreciate the enhancement.

Resolving.
